qiskit_experiments/calibration_management/calibration_utils.py

```diff
@@ -12,51 +12,171 @@
+from functools import lru_cache
+import re
+import retworkx as rx
 
+from qiskit.circuit import ParameterExpression, Parameter
+from qiskit.pulse import ScheduleBlock
 
-def used_in_calls(schedule_name: str, schedules: List[ScheduleBlock]) -> Set[str]:
-    """Find the schedules in the given list that call a given schedule by name.
+from qiskit_experiments.exceptions import CalibrationError
+from qiskit_experiments.calibration_management.calibration_key_types import ScheduleKey
+
+
+# The channel indices need to be parameterized following this regex.
+CHANNEL_PATTERN = r"^ch\d+(\.\d+)*(\$\d+){0,1}$"
+CHANNEL_PATTERN_REGEX = re.compile(CHANNEL_PATTERN)
+
+
+def update_schedule_dependency(schedule: ScheduleBlock, dag: rx.PyDiGraph, key: ScheduleKey):
+    """Update a DAG of schedule dependencies.
+
+    Args:
+        schedule: A ScheduleBlock that potentially has references to other schedules
+            that are already present in the dag.
+        dag: A directed acyclic graph that encodes schedule dependencies using references.
+        key: The schedule key which also contains the qubits.
+    """
+    # First, check if the node is already in the DAG.
+    try:
+        # If it already is in the DAG we remove the existing edges and add the new ones later.
+        parent_idx = dag.nodes().index(key)
+        for successor in dag.successors(parent_idx):
+            dag.remove_edge(parent_idx, dag.nodes().index(successor))
+
+    except ValueError:
+        # The schedule is not in the DAG: we add a new node.
+        parent_idx = dag.add_node(key)
+
+    for reference in schedule.references:
+        ref_key = ScheduleKey(reference[0], key.qubits)
+        dag.add_edge(parent_idx, _get_node_index(ref_key, dag), None)
+
+
+def used_in_references(keys: Set[ScheduleKey], dag: rx.PyDiGraph) -> Set[str]:
+    """Find all the schedules in the DAG that reference the given schedules.
 
     Args:
-        schedule_name: The name of the callee to identify.
-        schedules: A list of potential caller schedules to search.
+        keys: A list of schedules keys to which references may exist.
+        dag: The dag that represents the dependencies between schedule references.
 
     Returns:
-        A set of schedule names that call the given schedule.
+        A set of schedules that reference the given schedules.
     """
-    caller_names = set()
+    callers = set()
 
-    for schedule in schedules:
-        if _used_in_calls(schedule_name, schedule):
-            caller_names.add(schedule.name)
+    for key in keys:
+        callers.update(dag.nodes()[idx] for idx in rx.ancestors(dag, _get_node_index(key, dag)))
 
-    return caller_names
+    return set(key.schedule for key in callers)
 
 
-def _used_in_calls(schedule_name: str, schedule: ScheduleBlock) -> bool:
-    """Recursively find if the schedule calls a schedule with name ``schedule_name``.
+def _referred_by(key: ScheduleKey, dag: rx.PyDiGraph) -> Set[ScheduleKey]:
```

Comment:
Can the be removed now? Just a nitpick.

Reply:
Ah yes. Forgot to remove it thanks. See here: [32ab306](https://github.com/Qiskit/qiskit-experiments/pull/889/commits/32ab30631e901210a7b38cc51e1497b5eff6154c)